### PR TITLE
nixos: provide /etc/hostname

### PR DESCRIPTION
**Pull request by bjornfor**
###### Motivation for this change
/etc/hostname is the file used by hostnamectl(1) and the
org.freedesktop.hostname1 dbus service (both provided by systemd) to get
the "static hostname". Better provide it so that users of those
tools/services get a proper hostname.

An example of an issue created by the lack of /etc/hostname is that the
bluetooth stack on NixOS identifies itself to peers as "BlueZ $VERSION"
instead of the hostname.

References:
https://www.freedesktop.org/software/systemd/man/hostname.html

###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by bjornfor** [comment]
One side effect I just noticed. Without this PR, hostnamectl prints

```
Static hostname: n/a
Transient hostname: the_hostname
...
```

and if I run ```sudo hostname foo```, that "Transient hostname" line gets updated. As expected.

With this PR, only the "Static hostname" line is shown. (I guess that's fine, now that it includes the proper hostname.) BUT, there is no "Transient hostname" line, even after ```sudo hostname foo```. That surprised me.

cc @edolstra

**Comment by grahamc** [comment]
I added "policy discussion" since adding new files to /etc by default seems to require some discussion. Feel free to remove if you disagree :)

**Comment by Mic92** [comment]
Strange, why does `Bluez` not just call `hostname(2)` like any other tool?

**Comment by bjornfor** [comment]
@Mic92: I think it's because they are fetching more than just the "transient" hostname that you get from the kernel. Basically, they get "PrettyHostname", "StaticHostname" and "Chassis", all provided by org.freedesktop.hostname1[1].

See bluez code here: http://git.kernel.org/cgit/bluetooth/bluez.git/tree/plugins/hostname.c#n133

[1] https://www.freedesktop.org/wiki/Software/systemd/hostnamed/

---

I thought about patching bluez at first. But when I saw that hostnamectl prints "Static hostname: n/a" without an ```/etc/hostname``` file, I dropped that thought. (How much energy should we spend to work around/against upstream?)

Another option could be patching systemd to treat the hostname from the kernel as "static hostname". But I kind of like the idea of a static vs transient hostname. Snippet from hostnamed documentation:

> The transient (dynamic) host name is the one configured via the kernel's sethostbyname(). It can be different from the static hostname in case DHCP or mDNS have been configured to change the name based on network information.

**Comment by jokogr** [comment]
I use the `/etc/hostname` file to identify my machine and load a specific configuration file from my centralized configuration repository, e.g. have a look at https://git.joko.gr/joko/etc-nixos/src/master/configuration.nix#L4 (idea taken from @nathan7's configuration files).

Would this change cause any issues on my setups?

**Comment by bjornfor** [comment]
@jokogr: AFAICT, just change ```networking.hostName = builtins.readFile /etc/hostname;``` to ```networking.hostName = "thehostname";```.

**Comment by bjornfor** [comment]
> BUT, there is no "Transient hostname" line, even after sudo hostname foo. That surprised me.

It turns out that hostnamectl caches the "Transient hostname" for a few minutes (5?). Eventually it picks up the value from "sudo hostname foo".

**Comment by edolstra** [comment]
This looks fine to me.

**Comment by LnL7** [comment]
I thought we wanted as little mutable stuff in /etc as possible, unless it makes sense. (eg. a service that can reconfigure itself without restarting)
In this case the contents of /etc/hostname might not even match what is returned by `hostname(2)`.

**Comment by bjornfor** [comment]
> I thought we wanted as little mutable stuff in /etc as possible, unless it makes sense. (eg. a service that can reconfigure itself without restarting)

This PR creates an immutable /etc/hostname file, like most other files in /etc created by NixOS. (It'll be a symlink to /etc/static/hostname which in turn is a symlink to /nix/store/xybf7pjnnr1nq2j7apbx1110zqhx7qa1-etc-hostname.)

> In this case the contents of /etc/hostname might not even match what is returned by hostname(2).

Correct. That's why systemd has the concept of static vs transient hostname. The static name is stored in /etc/hostname and the transient (dynamic) hostname is the one stored in the kernel. See https://www.freedesktop.org/wiki/Software/systemd/hostnamed/.

**Comment by LnL7** [comment]
With mutable I was referring to the /etc/static symlink that changes with a nixos-rebuild. Anyway it looks fine to me, I just expected there would be more discussion around this.

File: nixos/modules/tasks/network-interfaces.nix [hunk 1]
```diff
@@ -910,6 +910,13 @@ in
         domainname "${cfg.domain}"
       '';
 
+    # static hostname configuration needed for hostnamectl and the
+    # org.freedesktop.hostname1 dbus service (both provided by systemd)
+    environment.etc."hostname" = mkIf (cfg.hostName != "")
+      {
+        text = cfg.hostName;
```
Comment (Mic92) [inline nixos/modules/tasks/network-interfaces.nix]: I would be nice to have a trailing newline in this file.

```bash
$ cat /etc/hostname
turingmachine$
```

Comment (bjornfor) [inline nixos/modules/tasks/network-interfaces.nix]: I agree. Fixed.

**Pull request merged**